### PR TITLE
feat: harmonise sc-element types

### PR DESCRIPTION
**Pull request by NikitaZotov**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -543,12 +543,12 @@ Add quintuple to your template.
 ![elements order in quintuple](./docs/images/quintuple.jpg)
 
 # ScType 
-This class is an abstraction over type in sc-memory. It consists of static properties with different types, such as `Link`, `NodeConst`, `NodeStruct`, `NodeVarTuple`, etc. with corresponding check functions `isNode`, `isEdge`, and others.
+This class is an abstraction over type in sc-memory. It consists of static properties with different types, such as `Link`, `NodeConst`, `NodeStructure`, `NodeVarTuple`, etc. with corresponding check functions `isNode`, `isConnector`, and others.
```
Comment (kilativ-dotcom) [inline README.md]:
```suggestion
This class is an abstraction over type in sc-memory. It consists of static properties with different types, such as `NodeLink`, `NodeConst`, `NodeStructure`, `NodeVarTuple`, etc. with corresponding check functions `isNode`, `isConnector`, and others.
```
mentioned types here are all nodes, add some connectors

File: docs/changelog.md [hunk 2]
```diff
@@ -40,7 +40,62 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - New sc-event types: `AfterGenerateConnector`, `AfterGenerateEdge`, `BeforeEraseConnector`, `BeforeEraseEdge` were added.
 
+- All sc-types were also redesigned to a common style. They were deprecated, new ones were added. See changes in the table below.
```
Comment (kilativ-dotcom) [inline docs/changelog.md]:
```suggestion
- All sc-types were redesigned to a common style. They were deprecated, new ones were added. See changes in the table below.
```

File: docs/changelog.md [hunk 3]
```diff
@@ -40,7 +40,62 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  | ScType.LinkVarClass           | ScType.VarNodeLinkClass   |
+  | ScType.NodeVarStruct          | ScType.VarNodeStructure   |
+  | ScType.NodeVarTuple           | ScType.VarNodeTuple       |
+  | ScType.NodeVarRole            | ScType.VarNodeRole        |
+  | ScType.NodeVarNoRole          | ScType.VarNodeNoRole      |
+  | ScType.NodeVarClass           | ScType.VarNodeClass       |
+  | ScType.NodeVarMaterial        | ScType.VarNodeMaterial    |
+
+  From now on, all sc-links are sc-nodes. Types of actual and inactual temporal membership sc-arc were added.
+  Also, all possible combinations of subtypes in sc-types have been added to the ScType.
```
Comment (kilativ-dotcom) [inline docs/changelog.md]:
```suggestion
  - All sc-links are sc-nodes.
  - Types of actual and inactual temporal membership sc-arc were added.
  - All possible combinations of subtypes in sc-types have been added to the ScType.
```

File: docs/changelog.md [hunk 4]
```diff
@@ -49,11 +104,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `ScEvent` class was renamed to `ScEventSubscription`
 - `ScEventParams` class was renamed to `ScEventSubscriptionParams`
 
+### Fixed
+- Checking of all syntactic and semantic subtypes for types in ScType `merge` method.
+- Now sc-link is sc-node
```
Comment (kilativ-dotcom) [inline docs/changelog.md:109]:
```suggestion
- sc-link is sc-node
```

File: src/ScClient.ts [hunk 5]
```diff
@@ -222,26 +222,26 @@ export class ScClient {
...
               type: cmd.type.value,
-              src: transformConnectorInfo(construction, cmd.data.src),
-              trg: transformConnectorInfo(construction, cmd.data.trg),
             };
           }
-          if (cmd.type.isLink()) {
+          else if (cmd.type.isConnector()) {
             return {
-              el: "link",
+              el: "edge",
```
Comment (kilativ-dotcom) [inline src/ScClient.ts:241]: shouldn't it be connector and not edge?

File: src/ScSet.ts [hunk 6]
```diff
@@ -168,7 +168,7 @@ export class ScSet {
     const templ: ScTemplate = new ScTemplate();
     templ.triple(
       this._addr,
-      [ScType.EdgeAccessVarPosPerm, "_edge"],
+      [ScType.VarPermPosArc, "_edge"],
```
Comment (kilativ-dotcom) [inline src/ScSet.ts]: not edge, replace everywhere

File: src/ScType.ts [hunk 7]
```diff
@@ -38,79 +54,116 @@ export class ScType {
...
+  }
+
+  public isVar(): boolean {
+    return (this._value & sc_type_var) != 0;
+  }
+
   public hasDirection(): boolean {
-    return (this._value & sc_type_uedge_common) == 0;
+    return (this._value & sc_type_common_edge) == 0;
   }
```
Comment (kilativ-dotcom) [inline src/ScType.ts]: shouldn't it just return isArc()? now all nodes are having direction

File: src/ScType.ts [hunk 8]
```diff
@@ -38,79 +54,116 @@ export class ScType {
...
     return (this._value & sc_type_node) != 0;
   }
 
+  public isLink(): boolean {
+    return (this._value & sc_type_node_link) == sc_type_node_link;
+  }
+
+  public isConnector(): boolean {
+    return (this._value & sc_type_connector) != 0;
+  }
```
Comment (kilativ-dotcom) [inline src/ScType.ts]: why are some methods comparing result with 0 and some are comparing result with type? also a lot of duplicated code, can be in some method that takes type and returns boolean

File: src/ScType.ts [hunk 9]
```diff
@@ -121,12 +174,94 @@ export class ScType {
...
-  public merge(other: ScType): ScType {
-    const t1 = this._value & sc_type_element_mask;
-    const t2 = other._value & sc_type_element_mask;
+  protected isNotCompatibleByMask(newType: ScType, mask: number): boolean {
+    const subtype = this.value & mask;
+    const newSubtype = newType.value & mask;
+    return subtype != sc_type_unknown && subtype != newSubtype;
+  }
+
+  protected isExpendableTo(newType: ScType): boolean { // it is equal to `sc_storage_is_type_expendable_to` in the sc-machine
```
Comment (kilativ-dotcom) [inline src/ScType.ts]: use extend instead of expand everywhere

File: src/ScType.ts [hunk 10]
```diff
@@ -121,12 +174,94 @@ export class ScType {
...
   }
 
-  public merge(other: ScType): ScType {
-    const t1 = this._value & sc_type_element_mask;
-    const t2 = other._value & sc_type_element_mask;
+  protected isNotCompatibleByMask(newType: ScType, mask: number): boolean {
+    const subtype = this.value & mask;
+    const newSubtype = newType.value & mask;
+    return subtype != sc_type_unknown && subtype != newSubtype;
+  }
```
Comment (kilativ-dotcom) [inline src/ScType.ts]: this method can be static that receives two ints and mask, no need to construct new ScType just to pass it here and get its value

File: src/ScType.ts [hunk 11]
```diff
@@ -121,12 +174,94 @@ export class ScType {
...
+      return false;
+    }
+
+    if (this.isLink()) {
+      if (!newType.isLink()) {
+        return false;
+      }
+
+      const thisType = new ScType(thisValue & ~sc_type_node_link);
+      newType = new ScType(newValue & ~sc_type_node_link);
```
Comment (kilativ-dotcom) [inline src/ScType.ts:204]: why newType is not const?

File: src/ScType.ts [hunk 12]
```diff
@@ -137,127 +272,329 @@ export class ScType {
...
+  static readonly ConstActualTempArc = new ScType(sc_type_const | sc_type_actual_arc | sc_type_temp_arc);
+  static readonly VarActualTempArc = new ScType(sc_type_var | sc_type_actual_arc | sc_type_temp_arc);
+  static readonly ConstInactualTempArc = new ScType(sc_type_const | sc_type_inactual_arc | sc_type_temp_arc);
+  static readonly VarInactualTempArc = new ScType(sc_type_var | sc_type_inactual_arc | sc_type_temp_arc);
+
+  // positivity
+  static readonly PosArc = new ScType(sc_type_pos_arc);
+  static readonly NegArc = new ScType(sc_type_neg_arc);
+
+  // fuzzy sc-arc
```
Comment (kilativ-dotcom) [inline src/ScType.ts]:
```suggestion
  // fuzziness
```

Comment (kilativ-dotcom) [reply src/ScType.ts]: and place it after // positive sc-arcs section to follow common style when new subtype is introduced and then it is combined with all possible types

File: src/ScType.ts [hunk 13]
```diff
@@ -137,127 +272,329 @@ export class ScType {
...
+  static readonly ConstFuzArc = new ScType(sc_type_const | sc_type_fuz_arc);
+  static readonly VarFuzArc = new ScType(sc_type_var | sc_type_fuz_arc);
+
+  // semantic sc-node types
+  static readonly NodeLink = new ScType(sc_type_node_link);
+  static readonly NodeLinkClass = new ScType(sc_type_node_link | sc_type_node_class);
+  static readonly NodeTuple = new ScType(sc_type_node_tuple);
+  static readonly NodeStructure = new ScType(sc_type_node_structure);
+  static readonly NodeRole = new ScType(sc_type_node_role);
+  static readonly NodeNoRole = new ScType(sc_type_node_norole);
```
Comment (kilativ-dotcom) [inline src/ScType.ts]:
```suggestion
  static readonly NodeNoRole = new ScType(sc_type_node_no_role);
```

File: src/constants.ts [hunk 14]
```diff
@@ -1,62 +1,56 @@
...
+export const sc_type_inactual_arc = (sc_type_membership_arc | 0x2000)
+
+// sc-arc permanence
+export const sc_type_temp_arc = (sc_type_membership_arc | 0x400)
+export const sc_type_perm_arc = (sc_type_membership_arc | sc_type_actual_arc | 0x800)
+
+// sc-arc positivity
+export const sc_type_pos_arc = (sc_type_membership_arc | 0x80)
+export const sc_type_neg_arc = (sc_type_membership_arc | 0x100)
+export const sc_type_fuz_arc = (sc_type_membership_arc | 0x200)
```
Comment (kilativ-dotcom) [inline src/constants.ts:27]: fuz arc in not positive, move it away from positive arcs for more obvious distinction

File: src/constants.ts [hunk 15]
```diff
@@ -1,62 +1,56 @@
...
+export const sc_type_node_material = (sc_type_node | 0x2000)
+
+// type mask
+export const sc_type_element_mask = (sc_type_node | sc_type_connector)
+export const sc_type_connector_mask = (sc_type_common_edge | sc_type_common_arc | sc_type_membership_arc)
+export const sc_type_arc_mask = (sc_type_common_arc | sc_type_membership_arc)
+
+export const sc_type_constancy_mask = (sc_type_const | sc_type_var)
+export const sc_type_actuality_mask = (sc_type_actual_arc | sc_type_inactual_arc)
+export const sc_type_permanency_mask = (sc_type_perm_arc | sc_type_temp_arc)
```
Comment (kilativ-dotcom) [inline src/constants.ts:46]: this one makes actuality a part of permanency, but inactuality is not part of permanency
just warning about inconsistency, probably sc-machine has same issue

File: src/constants.ts [hunk 16]
```diff
@@ -1,62 +1,56 @@
...
+export const sc_type_positivity_mask = (sc_type_pos_arc | sc_type_neg_arc | sc_type_fuz_arc)
```
Comment (kilativ-dotcom) [inline src/constants.ts]: fuz is not positivity

File: src/constants.ts [hunk 17]
```diff
@@ -1,62 +1,56 @@
...
+export const sc_type_membership_arc_mask = (sc_type_actuality_mask | sc_type_permanency_mask | sc_type_positivity_mask)
+export const sc_type_common_arc_mask = (sc_type_common_arc)
+export const sc_type_common_edge_mask = (sc_type_common_edge)
+
+export const sc_type_node_mask = (
+    sc_type_node_link | sc_type_node_tuple | sc_type_node_structure | sc_type_node_role | sc_type_node_norole
+    | sc_type_node_class | sc_type_node_superclass | sc_type_node_material)
+export const sc_type_node_link_mask = (sc_type_node | sc_type_node_link | sc_type_node_class)
 
 export const DEFAULT_KEYNODES_CASHE_SIZE = 5000;
```
Comment (kilativ-dotcom) [inline src/constants.ts]:
```suggestion
export const DEFAULT_KEYNODES_CACHE_SIZE = 5000;
```

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]

File: docs/changelog.md (hunk 4 above)
Comment (NikitaZotov) [reply docs/changelog.md:109]: No

File: src/ScClient.ts (hunk 5 above)
Comment (NikitaZotov) [reply src/ScClient.ts:241]: no, it is a part of sc-json

File: src/ScType.ts (hunk 10 above)
Comment (NikitaZotov) [reply src/ScType.ts]: what's the advantage?

File: src/ScType.ts (hunk 11 above)
Comment (NikitaZotov) [reply src/ScType.ts:204]: It is provided as argument

File: src/constants.ts (hunk 15 above)
Comment (NikitaZotov) [reply src/constants.ts:46]: all good here

File: docs/changelog.md [hunk 18]
```diff
@@ -40,7 +40,62 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  | ScType.NodeVarStruct          | ScType.VarNodeStructure   |
+  | ScType.NodeVarTuple           | ScType.VarNodeTuple       |
+  | ScType.NodeVarRole            | ScType.VarNodeRole        |
+  | ScType.NodeVarNoRole          | ScType.VarNodeNoRole      |
+  | ScType.NodeVarClass           | ScType.VarNodeClass       |
+  | ScType.NodeVarMaterial        | ScType.VarNodeMaterial    |
+
+- All sc-links are sc-nodes.
+- Types of actual and inactual temporal membership sc-arc were added.
+- Also, all possible combinations of subtypes in sc-types have been added to the ScType.
```
Comment (kilativ-dotcom) [inline docs/changelog.md]:
```suggestion
- All possible combinations of subtypes in sc-types have been added to the ScType.
```

File: src/ScType.ts (hunk 10 above)
Comment (kilativ-dotcom) [reply src/ScType.ts]: same as in py-sc-client

File: src/ScType.ts (hunk 12 above)
Comment (kilativ-dotcom) [reply src/ScType.ts]: and place it before line that starts with '// fuzzy sc-arcs'

File: src/constants.ts (hunk 15 above)
Comment (kilativ-dotcom) [reply src/constants.ts:46]: in checking type extendability for actualArc.isExtendableTo(temporalArc):
1. first arc is actual, second arc has no actuality, false isn't returned
2. first arc has no permanency(but comparing with this mask says that it has actual permanency), second arc has temporal permanency, false **is**(!) returned
but it should be allowed

File: src/constants.ts [hunk 19]
```diff
@@ -1,62 +1,58 @@
...
+export const sc_type_element_mask = (sc_type_node | sc_type_connector)
+export const sc_type_connector_mask = (sc_type_common_edge | sc_type_common_arc | sc_type_membership_arc)
+export const sc_type_arc_mask = (sc_type_common_arc | sc_type_membership_arc)
+
+export const sc_type_constancy_mask = (sc_type_const | sc_type_var)
+export const sc_type_actuality_mask = (sc_type_actual_arc | sc_type_inactual_arc)
+export const sc_type_permanency_mask = (sc_type_perm_arc | sc_type_temp_arc)
+export const sc_type_positivity_mask = (sc_type_pos_arc | sc_type_neg_arc)
+
+export const sc_type_membership_arc_mask = (sc_type_actuality_mask | sc_type_permanency_mask | sc_type_positivity_mask)
```
Comment (kilativ-dotcom) [inline src/constants.ts]: fuzzy arc is not in membership mask, is it ok?

File: docs/changelog.md (hunk 18 above)
Comment (NikitaZotov) [reply docs/changelog.md]: Okay

File: src/constants.ts (hunk 15 above)
Comment (NikitaZotov) [reply src/constants.ts:46]: Yes, there is a problem, but not with what you said. If the first arc is temporal without specified actuality and the second arc is actual, true isn't returned. I fixed this in sc-machine commit https://github.com/ostis-ai/sc-machine/pull/426/commits/059f5779ca972effbe10c406961271b2d38f2d77

File: docs/changelog.md (hunk 4 above)
Comment (NikitaZotov) [reply docs/changelog.md:109]: See py-sc-client

File: src/ScType.ts [hunk 20]
```diff
@@ -1,33 +1,49 @@
...
+  sc_type_actual_arc,
+  sc_type_inactual_arc,
+  sc_type_node_link,
   sc_type_node,
-  sc_type_node_abstract,
   sc_type_node_class,
+  sc_type_node_superclass,
   sc_type_node_material,
-  sc_type_node_norole,
+  sc_type_node_no_role,
```
Comment (kilativ-dotcom) [inline src/ScType.ts]: if no_role is splitted with underscore, shouldn't sc-machine follow same style?

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]
need to decide what to do with the method that checks extendability and  ci should say that everything is ok

File: src/ScType.ts (hunk 12 above)
Comment (NikitaZotov) [reply src/ScType.ts]: fixed

File: src/ScType.ts (hunk 10 above)
Comment (NikitaZotov) [reply src/ScType.ts]: fixed

File: src/ScType.ts (hunk 20 above)
Comment (NikitaZotov) [reply src/ScType.ts]: fixed

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]
fix ci

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]
and use non_role_relation

File: docs/changelog.md [hunk 21]
```diff
@@ -40,7 +40,62 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  | ScType.NodeConstRole          | ScType.ConstNodeRole      |
+  | ScType.NodeConstNonRole        | ScType.ConstNodeNonRole    |
+  | ScType.NodeConstClass         | ScType.ConstNodeClass     |
+  | ScType.NodeConstMaterial      | ScType.ConstNodeMaterial  |
+  | ScType.LinkVar                | ScType.VarNodeLink        |
+  | ScType.LinkVarClass           | ScType.VarNodeLinkClass   |
+  | ScType.NodeVarStruct          | ScType.VarNodeStructure   |
+  | ScType.NodeVarTuple           | ScType.VarNodeTuple       |
+  | ScType.NodeVarRole            | ScType.VarNodeRole        |
+  | ScType.NodeVarNonRole          | ScType.VarNodeNonRole      |
```
Comment (kilativ-dotcom) [inline docs/changelog.md]: non role weren't deprecated

File: docs/changelog.md [hunk 22]
```diff
@@ -49,11 +104,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Now sc-link is sc-node
+
 ### Deprecated
+- ScType methods: `isEdge`, `isStruct`
+- sc-types: `ScType.EdgeUCommon`, `ScType.EdgeDCommon`, `ScType.EdgeUCommonConst`, `ScType.EdgeDCommonConst`, `ScType.EdgeAccess`, `ScType.EdgeAccessConstPosPerm`, `ScType.EdgeAccessConstNegPerm`, `ScType.EdgeAccessConstFuzPerm`, `ScType.EdgeAccessConstPosTemp`, `ScType.EdgeAccessConstNegTemp`, `ScType.EdgeAccessConstFuzTemp`, `ScType.EdgeUCommonVar`, `ScType.EdgeDCommonVar`, `ScType.EdgeAccessVarPosPerm`, `ScType.EdgeAccessVarNegPerm`, `ScType.EdgeAccessVarFuzPerm`, `ScType.EdgeAccessVarPosTemp`, `ScType.EdgeAccessVarNegTemp`, `ScType.EdgeAccessVarFuzTemp`, `ScType.NodeConst`, `ScType.NodeVar`, `ScType.Link`, `ScType.LinkClass`, `ScType.NodeStruct`, `ScType.LinkConst`, `ScType.LinkConstClass`, `ScType.NodeConstTuple`, `ScType.NodeConstStruct`, `ScType.NodeConstRole`, `ScType.NodeConstNonRole`, `ScType.NodeConstClass`, `ScType.NodeConstMaterial`, `ScType.LinkVar`, `ScType.LinkVarClass`, `ScType.NodeVarStruct`, `ScType.NodeVarTuple`, `ScType.NodeVarRole`, `ScType.NodeVarNonRole`, `ScType.NodeVarClass`, `ScType.NodeVarMaterial`
```
Comment (kilativ-dotcom) [inline docs/changelog.md]: non role weren't deprecated

File: src/ScType.ts [hunk 23]
```diff
@@ -137,127 +280,328 @@ export class ScType {
...
+
+  /*!
+  * @deprecated NodeConstRole is deprecated. Use ConstNodeRole instead.
+  */
+  static readonly NodeConstRole = new ScType(ScType.ConstNodeRole);
+
+  /*!
+  * @deprecated NodeConstNonRole is deprecated. Use ConstNodeNonRole instead.
+  */
+  static readonly NodeConstNonRole = new ScType(ScType.ConstNodeNonRole);
```
Comment (kilativ-dotcom) [inline src/ScType.ts]:
```suggestion
  * @deprecated NodeConstNoRole is deprecated. Use ConstNodeNonRole instead.
  */
  static readonly NodeConstNoRole = new ScType(ScType.ConstNodeNonRole);
```

File: src/ScType.ts [hunk 24]
```diff
@@ -137,127 +280,328 @@ export class ScType {
...
+
+  /*!
+  * @deprecated NodeVarRole is deprecated. Use VarNodeRole instead.
+  */
+  static readonly NodeVarRole = new ScType(ScType.VarNodeRole);
+
+  /*!
+  * @deprecated NodeVarNonRole is deprecated. Use VarNodeNonRole instead.
+  */
+  static readonly NodeVarNonRole = new ScType(ScType.VarNodeNonRole);
```
Comment (kilativ-dotcom) [inline src/ScType.ts]:
```suggestion
  * @deprecated NodeVarNoRole is deprecated. Use VarNodeNonRole instead.
  */
  static readonly NodeVarNoRole = new ScType(ScType.VarNodeNonRole);
```

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]

**Review by kilativ-dotcom — APPROVED** [review]

**Pull request merged**